src/mongoose_rabbit_worker.erl

```diff
@@ -210,20 +208,28 @@ make_chat_msg(From, To, UserMsg) ->
 is_user_online(online) -> true;
 is_user_online(offline) -> false.
 
--spec publish_message_and_wait_for_confirm(Channel :: pid(),
-                                           Exchange :: binary(),
+-spec publish_message_and_wait_for_confirm(Exchange :: binary(),
                                            RoutingKey :: binary(),
-                                           Message :: binary()) -> term().
-publish_message_and_wait_for_confirm(Channel, Exchange, RoutingKey, Message) ->
+                                           Message :: binary(),
+                                           Opts :: publish_opts()) -> term().
+publish_message_and_wait_for_confirm(Exchange, RoutingKey, Message,
```

Comment:
The name should indicate that we *maybe* wait for confirm.

Reply:
Also when confirms are enabled the client code talking the the worker needs a way to get info when confirms have arrived. Otherwise there's no way to implement the backpressure mechanisms based on the confirms. So either with confirm on these are calls (or are always calls) or you somehow communicate back to the client.

Reply:
But how can we implement backpressure here? Isn't it enough to accumulate events to be sent in worker processeses message queues? 

As far as I understand `mod_event_pusher_rabbit:push_event/3` is called in c2s process context. Here is the flow:
1. c2s "emits" appropriate hooks (e.g. [user_send_packet hook])
2. Event pusher adds a handler for this hook ([code](https://github.com/esl/MongooseIM/blob/master/src/event_pusher/mod_event_pusher_hook_translator.erl#L34))
3. [The handler calls](https://github.com/esl/MongooseIM/blob/master/src/event_pusher/mod_event_pusher_hook_translator.erl#L65) `mod_event_pusher:push_event/3` which in turn [calls configured backends](https://github.com/esl/MongooseIM/blob/master/src/event_pusher/mod_event_pusher.erl#L40) (`mod_event_pusher_rabbit`) in our case.

Having this in mind doing a synchronous call to Rabbit workers we put pushing events on a critical path. Is that what we want? How can we implement the backpressure you mentioned? I think it requires a separate discussion and eventually a separate PR :)

[user_send_packet hook]: https://github.com/esl/MongooseIM/blob/beea15b37cadaf29b64b112970e0877898936778/src/ejabberd_c2s.erl#L877

Reply:
Ahh I see. I think it's fine then... I though that even if we wait for confirms too long the worker process can simply get overloaded by having tons of messages in it's mailbox. As C2S (or other client code) has no way of knowing of this overload condition. So that should be noted somehow, ideally prevented from.

Reply:
@mentels What's the final conclusion on that issue?

Reply:
Maybe for starters queue length monitoring would be enough? And that could be set to some value.

Reply:
Do you mean that a worker can monitor its queue length and drop requests if the value is too high?

Reply:
C2S monitors its queue and probably crash if it's too long. Yeah we loose messages that way but at least have a safety valve.